File: neetcode/python/22.generate-parentheses.py
#
# @lc app=leetcode id=22 lang=python3
#
# [22] Generate Parentheses
#

# @lc code=start
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        openers = n
        closers = n
        res = []

        def dfs(openers, closers, validParenthesis):
            if openers == 0 and closers == 0:
                res.append(validParenthesis)
                return 
            if openers > 0:
                dfs(openers - 1, closers, validParenthesis + '(')

            if closers > openers:
                dfs(openers, closers-1, validParenthesis + ')')

        dfs(openers, closers, '')
        return res
        
# @lc code=end

File: neetcode/go/22.generate-parentheses.go
/*
 * @lc app=leetcode id=22 lang=golang
 *
 * [22] Generate Parentheses
 */

// @lc code=start
func dfs(openers int, closers int, valid string, res *[]string) {
	if openers == 0 && closers == 0 {
		*res = append(*res, valid)
		return
	}

	if openers > 0 {
		dfs(openers-1, closers, valid+"(", res)
	}

	if closers > openers {
		dfs(openers, closers-1, valid+")", res)
	}
}

func generateParenthesis(n int) []string {
	openers := n
	closers := n
	var res []string

	dfs(openers, closers, "", &res)
	return res
}

// @lc code=end

